Bump to KCL v2.7.2

nunofernandes opened the pull request:
## Release notes
Upgrade Logstash Kinesis Input Plugin to AWS Kinesis Client Library (KCL) v2.7.2. This major version upgrade introduces enhanced compatibility with AWS SDK v2, improved LocalStack testing support, and more robust endpoint configuration.

## What does this PR do?

- Migrated from KCL v1.15.0 to v2.7.2
- Added support for custom Kinesis, DynamoDB, and CloudWatch endpoint configuration
- Updated AWS SDK dependencies to v2.33.0
- Refactored worker and record processing logic
- Enhanced error handling and logging
- Added comprehensive integration test infrastructure with LocalStack
- Improved proxy and credentials handling

## Why is it important/What is the impact to the user?

This PR fixes an impending EOL on the AWS KCL v1*.

## Checklist

- [X] My code follows the style guidelines of this project
- [ X] I have commented my code, particularly in hard-to-understand areas
- [ X] I have made corresponding changes to the documentation
- [ X] I have made corresponding change to the default configuration files (and/or docker env variables)
- [ X] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- Please review the extensive changes in worker and configuration handling
- Test with your specific Kinesis stream configurations
- Verify compatibility with existing Logstash deployments

## How to test this PR locally

You need to have docker compose and make.

```bash
make integration
```

## Related issues

- Closes #107
- Closes #108 

## Use cases

n/a

## Screenshots

n/a

## Logs

Been running that in the last 24 hours without issues. Log volume is +70k logs / minute.

## Important

This PR is partially generated by Gen AI tools.

nunofernandes committed:
chore: Setup development environment with JRuby and dependencies

This commit sets up the development environment for the
logstash-input-kinesis plugin. It includes:

- .envrc: Configures JRuby, Logstash paths, AWS credentials
  (optional), and adds vendor jars to the classpath.
- .ruby-version: Specifies JRuby version.
- Makefile: Defines common tasks (install, test, clean).
- Rakefile: Uses jar-dependencies to manage Java dependencies.

The Jars::Installer.new.vendor_jars! method is now called with
`write_require_file: false` to prevent writing a require file.

nunofernandes committed:
Upgrade Kinesis Input Plugin to AWS KCL v2.7.2

- Migrate to Kinesis Client Library (KCL) v2.7.2
- Add support for custom endpoint configuration
- Improve LocalStack and testing compatibility
- Update AWS SDK dependencies and configuration
- Enhance error handling and logging
- Refactor worker and record processing logic
- Add integration test infrastructure

Update to 2.7.2

nunofernandes commented:
> ❌ Author of the following commits did not sign a [Contributor Agreement](https://www.elastic.co/contributor-agreement): [ff259f4](https://github.com/logstash-plugins/logstash-input-kinesis/commit/ff259f4848309395e56d0d80d48b5549f6c63890), [b48d78b](https://github.com/logstash-plugins/logstash-input-kinesis/commit/b48d78be1a7ad0a32810014db6d14bf470cebcbe)
> 
> Please, read and sign the above mentioned agreement if you want to contribute to this project

Done!

nunofernandes commented:
Can't see any of the travis CI logs.. so :shrug:

javier-rodriguez-n26 commented:
> Can't see any of the travis CI logs.. so 🤷

Both jobs failed with the same error, look like there's something broken in the CI
```
env file /home/travis/build/logstash-plugins/logstash-input-kinesis/.ci/dockerjdk17.env not found: stat /home/travis/build/logstash-plugins/logstash-input-kinesis/.ci/dockerjdk17.env: no such file or directory
```

felixho1 commented:
May I ask, how to run the bundle to get the gem generated ?
First I ran `bundle install` successfully

```
bundle install

Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...
Fetching tilt 2.7.0
Fetching net-http 0.9.1
Fetching io-console 0.8.2 (java)
Fetching multi_json 1.19.1
Fetching kramdown 2.5.2
Fetching json 2.18.0 (java)
Fetching i18n 1.14.8
Fetching tzinfo-data 1.2025.3
Fetching ffi 1.17.3 (java)
Installing tilt 2.7.0
Installing multi_json 1.19.1
Installing i18n 1.14.8
Installing net-http 0.9.1
Installing tzinfo-data 1.2025.3
Installing io-console 0.8.2 (java)
Fetching reline 0.6.3
Installing json 2.18.0 (java)
Installing kramdown 2.5.2
Installing ffi 1.17.3 (java)
Installing reline 0.6.3
Fetching pry 0.16.0 (java)
Installing pry 0.16.0 (java)
Bundle complete! 6 Gemfile dependencies, 72 gems now installed.
Bundled gems are installed into `/Users/hofe/apps/elk/logstash-8.15.0/vendor/bundle/jruby`
```

Then I ran `bundle exec rake vendor` but failed with error `Gem::LoadError: You have already activated jar-dependencies 0.4.1, but your Gemfile requires jar-dependencies 0.5.5. Since jar-dependencies is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports jar-dependencies as a default gem.`

I did try AI to try help me, but still no luck.


This is the complete log
```
bundle exec rake vendor
bundler: failed to load command: rake (/Users/hofe/apps/elk/logstash-8.15.0/vendor/bundle/jruby/jruby/3.1.0/bin/rake)
Gem::LoadError: You have already activated jar-dependencies 0.4.1, but your Gemfile requires jar-dependencies 0.5.5. Since jar-dependencies is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports jar-dependencies as a default gem.
  check_for_activated_spec! at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:317
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:25
                       each at org/jruby/RubyArray.java:1981
                       each at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/spec_set.rb:229
                        map at org/jruby/RubyEnumerable.java:818
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:24
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler.rb:167
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/setup.rb:16
                 with_level at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/ui/shell.rb:173
                    silence at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/ui/shell.rb:119
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/setup.rb:16
                    require at org/jruby/RubyKernel.java:1184
           require_relative at org/jruby/RubyKernel.java:1213
                kernel_load at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli/exec.rb:57
                        run at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli/exec.rb:23
                       exec at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:452
                        run at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/command.rb:28
             invoke_command at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/invocation.rb:127
                   dispatch at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor.rb:538
                   dispatch at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:35
                      start at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/base.rb:584
                      start at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:29
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/exe/bundle:28
       with_friendly_errors at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/friendly_errors.rb:117
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/exe/bundle:20
                       load at org/jruby/RubyKernel.java:1220
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/bin/bundle:25
hofe@TAB118719 logstash-plugins/logstash-input-kinesis (nuno/issue-107-Kinesis-Client-Library-version-1.x-end-of-support) » bundle exec rspec                                                                                                  1 ↵
bundler: failed to load command: rspec (/Users/hofe/apps/elk/logstash-8.15.0/vendor/bundle/jruby/jruby/3.1.0/bin/rspec)
Gem::LoadError: You have already activated jar-dependencies 0.4.1, but your Gemfile requires jar-dependencies 0.5.5. Since jar-dependencies is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports jar-dependencies as a default gem.
  check_for_activated_spec! at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:317
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:25
                       each at org/jruby/RubyArray.java:1981
                       each at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/spec_set.rb:229
                        map at org/jruby/RubyEnumerable.java:818
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:24
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler.rb:167
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/setup.rb:16
                 with_level at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/ui/shell.rb:173
                    silence at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/ui/shell.rb:119
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/setup.rb:16
                    require at org/jruby/RubyKernel.java:1184
           require_relative at org/jruby/RubyKernel.java:1213
                kernel_load at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli/exec.rb:57
                        run at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli/exec.rb:23
                       exec at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:452
                        run at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/command.rb:28
             invoke_command at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/invocation.rb:127
                   dispatch at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor.rb:538
                   dispatch at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:35
                      start at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/base.rb:584
                      start at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:29
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/exe/bundle:28
       with_friendly_errors at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/friendly_errors.rb:117
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/exe/bundle:20
                       load at org/jruby/RubyKernel.java:1220
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/bin/bundle:25
hofe@TAB118719 logstash-plugins/logstash-input-kinesis (nuno/issue-107-Kinesis-Client-Library-version-1.x-end-of-support) » bundle exec rspec spec/inputs/kinesis_spec.rb                                                                      1 ↵
bundler: failed to load command: rspec (/Users/hofe/apps/elk/logstash-8.15.0/vendor/bundle/jruby/jruby/3.1.0/bin/rspec)
Gem::LoadError: You have already activated jar-dependencies 0.4.1, but your Gemfile requires jar-dependencies 0.5.5. Since jar-dependencies is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports jar-dependencies as a default gem.
  check_for_activated_spec! at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:317
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:25
                       each at org/jruby/RubyArray.java:1981
                       each at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/spec_set.rb:229
                        map at org/jruby/RubyEnumerable.java:818
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/runtime.rb:24
                      setup at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler.rb:167
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/setup.rb:16
                 with_level at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/ui/shell.rb:173
                    silence at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/ui/shell.rb:119
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/setup.rb:16
                    require at org/jruby/RubyKernel.java:1184
           require_relative at org/jruby/RubyKernel.java:1213
                kernel_load at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli/exec.rb:57
                        run at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli/exec.rb:23
                       exec at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:452
                        run at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/command.rb:28
             invoke_command at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/invocation.rb:127
                   dispatch at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor.rb:538
                   dispatch at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:35
                      start at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/vendor/thor/lib/thor/base.rb:584
                      start at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/cli.rb:29
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/exe/bundle:28
       with_friendly_errors at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/lib/bundler/friendly_errors.rb:117
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/lib/ruby/gems/shared/gems/bundler-2.6.9/exe/bundle:20
                       load at org/jruby/RubyKernel.java:1220
                     <main> at /Users/hofe/apps/elk/logstash-8.15.0/vendor/jruby/bin/bundle:25
```

nunofernandes commented:
Hello @felixho1 ,

I use fedora locally + logstash installed via rpm + direnv (that creates the required environment for building it). That said, I also included a Dockerfile so that the build process can be done independently of the "outside" environment. For that, you just need to have docker available and run:

```bash
$ docker build --target builder-kinesis --tag builder-kinesis .
$ docker run -ti -v $(pwd)/output:/output --entrypoint bash builder-kinesis -c "cp /build/logstash-input-kinesis-*.gem /output/"
$ ls -lah output/*gem
-rw-r--r--. 1 root root 66M Jan 24 15:14 output/logstash-input-kinesis-3.0.0-java.gem
```

felixho1 commented:
@nunofernandes Thank you so much for your help raising the PR. And thank you for the tips to build with docker.
I have tested locally on my machine, it works perfectly! Thank you once again.
I hope this PR get merged as soon as possible.

akash1810 commented:
Hi there, is there any update on getting this change released? I'm not a maintainer, but happy where I can.

jorgeazevedo commented on lib/logstash/inputs/kinesis.rb:
```diff
@@ -103,104 +112,175 @@ def register
+    @additional_settings.each do |key, value|
+      fn = "#{key}"
+      begin
+        configsBuilder.send(fn, value)
```
`ConfigsBuilder` does not have `withX()` methods, so this code always fails. This means that support for [passing additional settings](https://github.com/logstash-plugins/logstash-input-kinesis?tab=readme-ov-file#additional-kcl-settings) is effectively broken.

nunofernandes replied:
Was using it like this: https://github.com/logstash-plugins/logstash-input-kinesis/pull/109/changes#diff-191c9828dad651802568f81a6b6e1999060075bdc5bca55b0003451d70146e26R8

but let me see if I can fix that.

nunofernandes replied:
Would it make sense to move also the *_endpoint configs into the `additional_settings` basically this?
<img width="1443" height="561" alt="image" src="https://github.com/user-attachments/assets/48d97ca7-a181-43ea-a6bb-c5d6112faca3" />

or should be "outside" like they were? Thanks for any feedback

jorgeazevedo replied:
Gosh, it's tricky. `additional_settings` was a way to directly access all the `with` methods in the KinesisClientLibConfiguration form kclv1. For example, setting `initial_lease_table_read_capacity` was a way to call  [withInitialLeaseTableReadCapacity](https://github.com/awslabs/amazon-kinesis-client/blob/f1d60ec1a6c9eb5af2a0593ab814531fdd730989/src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java#L1250-L1263).

In KCL v2 this interface no longer exists, configuration is now split across multiple different objects. My best guess is that instead of a single `additional_settings` property, it would make more sense to have multiple properties, one for each object used by KCL v2 (e.g. `retrieval_config`, `checkpoint_config` etc)

In that light, I think the answer to your question is that the endpoint configuration should not be merged in as it does not related to setting the KCL config objects.

nunofernandes replied:
Ok.. I just pushed another commit to add those settings.. 

<img width="1449" height="391" alt="image" src="https://github.com/user-attachments/assets/8a23e381-e088-4b2b-90ed-bc9966f6d188" />

Let me know if that works.

nunofernandes committed:
Fix additional_settings to work with KCL v2 sub-config objects

KCL v2's ConfigsBuilder does not have withX() methods like KCL v1's
KinesisClientLibConfiguration. Settings now iterate through all sub-config
objects (CheckpointConfig, CoordinatorConfig, LeaseManagementConfig, etc.)
to find the correct target for each setting.

Also adds integration-clean Makefile target and exercises additional_settings
in integration tests.

mashhurs reviewed (changes requested):
@nunofernandes thank you for contribution. Great work.

I would like to have this separated:
- to upgrade the dependencies first
- add additional configs later with the use-cases. It looks to me introducing an endpoint only for kinesis makes sense as plugin name indicates it is for kinesis purpose only. I wonder if cloudwatch works with aws-integration plugin - https://www.elastic.co/docs/reference/logstash/plugins/plugins-inputs-cloudwatch?

Also, it looks like all sources are generated by AI tools. Under the PR description, I would ask to add an note, something like:

> [!IMPORTANT]
> This PR is partially/fully generated by Gen AI tools.

nunofernandes commented:
@mashhurs thanks for the review and the feedback!                                                                                                          
                                                                                                                                                             
1.  Splitting the PR                                                                                           
                                                              
  I understand the concern about PR size. However, the endpoint configs (kinesis_endpoint, dynamodb_endpoint, cloudwatch_endpoint) are not really a separate feature - they're part of the core KCL v2 client construction. In v2, the plugin must construct KinesisAsyncClient, DynamoDbAsyncClient, and CloudWatchAsyncClient individually and pass  them to ConfigsBuilder. The endpoint overrides are set on those builders, so they're inherently part of the migration, not an addition on top.

What I can split out is the integration test infrastructure (Dockerfile, docker-compose, LocalStack scripts, Makefile targets) into a follow-up PR. That would significantly reduce the diff here. Would that work for you?

2.  CloudWatch endpoint

The cloudwatch_endpoint here is not related to the logstash-input-cloudwatch plugin or the AWS integration plugin. KCL itself uses a CloudWatch client internally to publish worker/shard metrics (when `metrics => "cloudwatch"` is enabled). The endpoint override is needed so KCL's metrics reporting can work in environments with custom endpoints (VPC endpoints, LocalStack, etc.). It's a KCL dependency, not a plugin feature.

3.  AI disclosure

Updated the PR.


Hope that helps..

jorgeazevedo commented:
@mashhurs are you able to tell us if your team at elastic is planning to upgrade this plugin to kcl v2 and maintaining it in the future? I get the sense that this project is currently low priority, but it would be helpful to know if prioritising this work is in the horizon at all or not.

mashhurs commented:
@nunofernandes , @jorgeazevedo I am so sorry for slow review.
While reviewing I noticed that 
- plugin build was broken due to `jar-dependencies` (via `ruby-maven`) wasn't able to resolve dependencies properly
- our CI matrix was old/broken
I have prepared to fix them first, will accelerate landing - https://github.com/logstash-plugins/logstash-input-kinesis/pull/110

With your PR, I am thinking to separate the Integration tests. It is because so many unnecessary files/context introduced by AI (I might comment alot now) that we can easily land the main/core logic with following file changes only:
- modified:   lib/logstash/inputs/kinesis.rb
- modified:   lib/logstash/inputs/kinesis/worker.rb
- modified:   logstash-input-kinesis.gemspec
- modified:   spec/inputs/kinesis/worker_spec.rb
- modified:   spec/inputs/kinesis_spec.rb

We can add integration test just after landing our main/core functionality. I believe this accelerates our process so fast. WDYT?

mashhurs commented:
@nunofernandes I have raised a PR - https://github.com/logstash-plugins/logstash-input-kinesis/pull/111 based on your changes and I have included you as a co-author (thank you alot for contribution). The differences from your PR are 
- backward compatibility: instead introducing new endpoints we use them from `additional_settings` (where users were using)
- build issue fixes: build fails with jruby 10+ versions, added fixes in Rakefile
- more specs to make sure we don't break bwc and keep the functionalities safe


So, I would suggest closing this PR and add your integration tests once #111 gets merged.

robbavey commented:
Superseded by https://github.com/logstash-plugins/logstash-input-kinesis/pull/111, closing
